File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
import heapq
input=sys.stdin.readline

V,E=map(int,input().split())
graph={}

K=int(input())
l=[]
for i in range(E):
  u,v,w=map(int,input().split())
  if u in graph:
    if v in graph[u]:
      if graph[u][v]>w:
        graph[u][v]=w
    else:
      graph[u][v]=w
  else:
    graph[u]={v:w}


q=[]
if K in graph:
  for node in graph[K]:
    if graph[K][node]!=0 and graph[K][node]!=-1:
      heapq.heappush(q,[graph[K][node],node])

answer=[-1 for i in range(V+1)]
answer[K]=0
while q:
  cost,node=heapq.heappop(q)
  if answer[node]==-1 or answer[node]>cost:
    answer[node]=cost
    if node in graph:
      for nextNode in graph[node]:
        heapq.heappush(q,[graph[node][nextNode]+cost,nextNode])

for i in range(1,V+1):
  if i==K:
    print(0)
  elif answer[i]==-1:
    print('INF')
  else:
    print(answer[i])

File: 백준/Gold/1753. 최단경로/최단경로.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

function isSmall(a,b){
  if (a[1]<b[1]) return true
  return false
}

function heappush(q,item){
  q.push(item)
  let now=q.length-1
  let parent=Math.floor(now/2)
  while(parent>=1){
    if (isSmall(q[now],q[parent])){
      [q[now],q[parent]]=[q[parent],q[now]]
      now=parent
      parent=Math.floor(parent/2)
    } else break
  }
}

function heappop(q){
  const item=[...q[1]]
  q.splice(1,1)
  let now=1
  while(true){
    const left=now*2
    const right=now*2+1
    if (q.length<=left) break
    if (q.length<=right){
      if (isSmall(q[left],q[now])){
        [q[now],q[left]]=[q[left],q[now]]
        now=left
        continue
      }
      break
    }
    const Min=isSmall(q[left],q[right])?left:right
    if (isSmall(q[Min],q[now])){
      [q[now],q[Min]]=[q[Min],q[now]]
      now=Min
      continue
    } else break
  }
  return item
}

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const [V,E]=input[iter++].split(' ').map(Number)
  const K=Number(input[iter++])

  const graph={}

  for(let i=0;i<E;i++){
    const [u,v,w]=input[iter++].split(' ').map(Number)
    if (u in graph){
      if (v in graph[u]){
        if (graph[u][v]>w) graph[u][v]=w
      } else graph[u][v]=w
    } else {
      graph[u]={[v]:w}
    }
  }

  const answer=[...new Array(V+1)].map(item=>'INF')
  answer[K]=0

  const q=[0]
  const arr=Object.keys(graph[K]).map(item=>[item,graph[K][item]])
  for(let i=0;i<arr.length;i++){
    heappush(q,arr[i])
    const [nextnode,cost]=arr[i]
    if (answer[nextnode]==='INF') answer[nextnode]=cost
  }

  while(q.length>1){
    const [node,cost]=heappop(q)
    const nextkeys=node in graph?Object.keys(graph[node]):[]
    for(let i=0;i<nextkeys.length;i++){
      const nextnode=nextkeys[i]
      if (answer[nextnode]==='INF' || answer[nextnode]>cost+graph[node][nextnode]){
        answer[nextnode]=cost+graph[node][nextnode]
        heappush(q,[nextnode,cost+graph[node][nextnode]])
      }
    }
  }

  answer.shift()
  console.log(answer.join('\n'))
  process.exit()
})
